### PR TITLE
Adds Cis2Client

### PR DESCRIPTION
**Pull request by parv0888**
## Purpose

Adds Cis2 Client to address https://github.com/Concordium/concordium-rust-smart-contracts/issues/295

## Changes

Adds Cis2 Client to address https://github.com/Concordium/concordium-rust-smart-contracts/issues/295

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: concordium-cis2/src/cis2_client.rs [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+    // calls the `transfer` entrypoint of the CIS2 contract to transfer the given
+    // amount of tokens from the given owner to the given receiver.
+    // If the transfer is successful, it returns `Ok(())`, else it returns an `Err`.
+    pub fn transfer<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount + Copy>(
+        host: &mut impl HasHost<State, StateApiType = S>,
+        cis_contract_address: &ContractAddress,
+        token_id: T,
+        amount: A,
+        from: Address,
+        to: Receiver,
```
Comment (limemloh) [inline concordium-cis2/src/cis2_client.rs:158]: I think it is better to take the transfer directly here
```suggestion
        transfer: Transfer<T, A>,
```

File: concordium-cis2/src/cis2_client.rs [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+        // error. Since for a single query the response should be non-empty.
+        let balance = *balances.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
+
+        Ok(balance)
+    }
+
+    // calls the `transfer` entrypoint of the CIS2 contract to transfer the given
+    // amount of tokens from the given owner to the given receiver.
+    // If the transfer is successful, it returns `Ok(())`, else it returns an `Err`.
+    pub fn transfer<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount + Copy>(
```
Comment (limemloh) [inline concordium-cis2/src/cis2_client.rs:152]: I don't think you need Copy here
```suggestion
    pub fn transfer<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount>(
```

File: concordium-cis2/src/cis2_client.rs [hunk 3]
```diff
@@ -0,0 +1,433 @@
+//! CIS2 client is the intermediatory layer between any contract and
+//! CIS2 comliant contract.
+//!
+//! # Description
+//! It allows the contract to abstract away the logic of calling the
+//! CIS2 contract for the following methods
+//! - `supports_cis2` : Calls [`supports`](https://proposals.concordium.software/CIS/cis-0.html#supports)
+//! - `operator_of` : Calls [`operatorOf`](https://proposals.concordium.software/CIS/cis-2.html#operatorof)
+//! - `balance_of` : Calls [`balanceOf`](https://proposals.concordium.software/CIS/cis-2.html#balanceof)
+//! - `transfer` : Calls [`transfer`](https://proposals.concordium.software/CIS/cis-2.html#transfer)
```
Comment (limemloh) [inline concordium-cis2/src/cis2_client.rs:10]: A small example of how to set up a client and call `balance_of` and `transfer` would be nice here.

Comment (limemloh) [inline concordium-cis2/src/cis2_client.rs:10]: No implementation of `updateOperator`?

**Review by limemloh — CHANGES_REQUESTED** [review]
Overall good!
Have some comments and some suggestions for you to look at

File: concordium-cis2/src/cis2_client.rs [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+const SUPPORTS_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("supports");
+const OPERATOR_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("operatorOf");
+const BALANCE_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("balanceOf");
+const TRANSFER_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("transfer");
+
+/// Errors which can be returned by the `Cis2Client`.
+#[derive(Serialize, Debug, PartialEq, Eq, Reject)]
+pub enum Cis2ClientError {
+    /// When there is an error invoking the CIS2 contract.
+    InvokeContractError,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:26]: done

File: concordium-cis2/src/cis2_client.rs [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+    /// When there is an error parsing the parameters.
+    ParseParams,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:28]: done

File: concordium-cis2/src/cis2_client.rs [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+    /// When there is an error parsing the result.
+    ParseResult,
+}
+
+impl<T> From<CallContractError<T>> for Cis2ClientError {
+    fn from(_: CallContractError<T>) -> Self { Cis2ClientError::InvokeContractError }
+}
+
+impl From<ParseError> for Cis2ClientError {
+    fn from(_: ParseError) -> Self { Cis2ClientError::ParseParams }
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:38]: done

File: concordium-cis2/src/cis2_client.rs [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+pub struct Cis2Client;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:41]: done

File: concordium-cis2/src/cis2_client.rs [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+    // contract supports CIS2 standard. If the contract supports CIS2 standard,
+    // it returns the contract address, else it returns None.
+    pub fn supports_cis2<State, S: HasStateApi>(
+        host: &mut impl HasHost<State, StateApiType = S>,
+        cis_contract_address: &ContractAddress,
+    ) -> Result<Option<ContractAddress>, Cis2ClientError> {
+        let params = SupportsQueryParams {
+            queries: vec![StandardIdentifierOwned::new_unchecked(
+                CIS2_STANDARD_IDENTIFIER_STR.to_string(),
+            )],
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:54]: done

File: concordium-cis2/src/cis2_client.rs [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+        };
+
+        let parsed_res = match host.invoke_contract_read_only(
+            cis_contract_address,
+            &params,
+            SUPPORTS_ENTRYPOINT_NAME,
+            Amount::from_ccd(0),
+        )? {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            None => bail!(Cis2ClientError::InvokeContractError),
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:64]: done

File: concordium-cis2/src/cis2_client.rs [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+            .ok_or(Cis2ClientError::InvokeContractError)?;
+
+        Ok(supports_cis2.copied())
+    }
+
+    // calls the `operatorOf` entrypoint of the CIS2 contract to check if the given
+    // owner is an operator of the given contract. If the owner is an operator
+    // of the given contract, it returns true, else it returns false.
+    pub fn operator_of<State, S: HasStateApi>(
+        host: &mut impl HasHost<State, StateApiType = S>,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:85]: done

File: concordium-cis2/src/cis2_client.rs [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+pub struct Cis2Client;
+
+impl Cis2Client {
+    // calls the `supports` entrypoint of the CIS2 contract to check if the given
+    // contract supports CIS2 standard. If the contract supports CIS2 standard,
+    // it returns the contract address, else it returns None.
+    pub fn supports_cis2<State, S: HasStateApi>(
+        host: &mut impl HasHost<State, StateApiType = S>,
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:48]: done

File: concordium-cis2/src/cis2_client.rs [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+
+        let supports_cis2 = parsed_res
+            .results
+            .first()
+            .map(|f| match f {
+                SupportResult::NoSupport => Option::None,
+                SupportResult::Support => Option::Some(cis_contract_address),
+                SupportResult::SupportBy(contracts) => contracts.first(),
+            })
+            .ok_or(Cis2ClientError::InvokeContractError)?;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:76]: done

File: concordium-cis2/src/cis2_client.rs [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+pub struct Cis2Client;
+
+impl Cis2Client {
+    // calls the `supports` entrypoint of the CIS2 contract to check if the given
+    // contract supports CIS2 standard. If the contract supports CIS2 standard,
+    // it returns the contract address, else it returns None.
+    pub fn supports_cis2<State, S: HasStateApi>(
+        host: &mut impl HasHost<State, StateApiType = S>,
+        cis_contract_address: &ContractAddress,
+    ) -> Result<Option<ContractAddress>, Cis2ClientError> {
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:50]: done

File: concordium-cis2/src/cis2_client.rs [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+        };
+
+        let is_operators = match host.invoke_contract_read_only(
+            cis_contract_address,
+            params,
+            OPERATOR_OF_ENTRYPOINT_NAME,
+            Amount::from_ccd(0),
+        )? {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            None => bail!(Cis2ClientError::InvokeContractError),
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:104]: done

File: concordium-cis2/src/cis2_client.rs [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+            Some(mut res) => OperatorOfQueryResponse::deserial(&mut res)?,
+        };
+
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        let is_operator = *is_operators.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:110]: done

File: concordium-cis2/src/cis2_client.rs (hunk 2 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:152]: done

File: concordium-cis2/src/cis2_client.rs (hunk 1 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:158]: done

File: concordium-cis2/src/cis2_client.rs [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+            Amount::from_ccd(0),
+        )? {
+            // Since the contract should return a response. If it doesn't, it is an error.
+            None => bail!(Cis2ClientError::InvokeContractError),
+            Some(mut res) => BalanceOfQueryResponse::<A>::deserial(&mut res)?,
+        };
+
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        let balance = *balances.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:144]: done

File: concordium-cis2/src/cis2_client.rs [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+        // error. Since for a single query the response should be non-empty.
+        let is_operator = *is_operators.0.first().ok_or(Cis2ClientError::InvokeContractError)?;
+
+        Ok(is_operator)
+    }
+
+    // calls the `balanceOf` entrypoint of the CIS2 contract to get the balance of
+    // the given owner for the given token. Returns the balance of the owner for
+    // the given token.
+    pub fn balance_of<State, S: HasStateApi, T: IsTokenId, A: IsTokenAmount + Copy>(
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:118]: done

File: concordium-cis2/CHANGELOG.md [hunk 18]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased changes
 
+## concordium-cis2 4.0.1 (2023-06-20)
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/CHANGELOG.md:5]: done

File: concordium-cis2/src/cis2_client.rs (hunk 3 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:10]: added

File: concordium-cis2/Cargo.toml [hunk 19]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "concordium-cis2"
-version = "4.0.0"
+version = "4.0.1"
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/Cargo.toml:3]: removed

File: concordium-cis2/src/cis2_client.rs (hunk 3 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:10]: done

File: concordium-cis2/src/cis2_client.rs [hunk 20]
```diff
@@ -0,0 +1,724 @@
...
+    /// When the response is invalid. Ex. When the response is empty vector for
+    /// a single query.
+    InvalidResponse,
+}
+
+impl<T: Serial> Serial for Cis2ClientError<T> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        match self {
+            Cis2ClientError::InvokeContractError(e) => {
+                out.write_u8(3)?;
```
Comment (limemloh) [inline concordium-cis2/src/cis2_client.rs:73]: We are not using the tag `2` for serializing any of the other variants, so I would prefer to use this here
```suggestion
                out.write_u8(2)?;
```

**Review by limemloh — CHANGES_REQUESTED** [review]
This looks great!
Only a few minor comments from me for this to be merged.

File: concordium-cis2/src/cis2_client.rs (hunk 20 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:73]: done

File: concordium-cis2/src/cis2_client.rs [hunk 21]
```diff
@@ -0,0 +1,724 @@
...
+    ) -> Result<bool, Cis2ClientError<E>> {
+        let params = TransferParams(vec![transfer]);
+        let (state_modified, _): (bool, Option<()>) =
+            self.invoke_contract(host, TRANSFER_ENTRYPOINT_NAME, &params)?;
+
+        Ok(state_modified)
+    }
+
+    /// Calls the `updateOperator` of the CIS2 contract.
+    /// If the update is successful, it returns `Ok(())`, else it returns an
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:335]: done

File: concordium-cis2/src/cis2_client.rs [hunk 22]
```diff
@@ -0,0 +1,724 @@
...
+        };
+
+        let mut res: BalanceOfQueryResponse<A> =
+            self.invoke_contract_read_only(host, BALANCE_OF_ENTRYPOINT_NAME, &params)?;
+        Cis2Client::first(&mut res.0)
+    }
+
+    /// Calls the `transfer` entrypoint of the CIS2 contract to transfer the
+    /// given amount of tokens from the given owner to the given receiver.
+    /// If the transfer is successful, it returns `Ok(())`, else it returns an
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:297]: done

File: concordium-cis2/src/cis2_client.rs [hunk 23]
```diff
@@ -0,0 +1,724 @@
...
+const OPERATOR_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("operatorOf");
+const BALANCE_OF_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("balanceOf");
+const TRANSFER_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("transfer");
+const UPDATE_OPERATOR_ENTRYPOINT_NAME: EntrypointName =
+    EntrypointName::new_unchecked("updateOperator");
+
+/// Wrapper for `Cis2Error<T>`. This allows for various traits to be
+/// implemented.
+#[derive(Debug)]
+pub struct Cis2ErrorWrapper<T>(Cis2Error<T>);
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:26]: done

File: concordium-cis2/src/cis2_client.rs [hunk 24]
```diff
@@ -0,0 +1,696 @@
...
+const TRANSFER_ENTRYPOINT_NAME: EntrypointName = EntrypointName::new_unchecked("transfer");
+const UPDATE_OPERATOR_ENTRYPOINT_NAME: EntrypointName =
+    EntrypointName::new_unchecked("updateOperator");
+
+pub type InvokeContractError<T> = CallContractError<Cis2Error<T>>;
+
+/// Errors which can be returned by the `Cis2Client`.
+#[derive(Debug)]
+pub enum Cis2ClientError<T> {
+    /// When there is an error invoking the CIS2 contract.
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:28]:
```suggestion
    /// Invoking the contract returned the given error.
```

File: concordium-cis2/src/cis2_client.rs [hunk 25]
```diff
@@ -0,0 +1,696 @@
...
+    InvokeContractError(InvokeContractError<T>),
+    /// When there is an error parsing the result.
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:30]:
```suggestion
    /// The response from the contract could not be parsed.
```

File: concordium-cis2/src/cis2_client.rs [hunk 26]
```diff
@@ -0,0 +1,696 @@
...
+    ParseResult,
+    /// When the response is invalid. Ex. When the response is empty vector for
+    /// a single query.
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:33]:
```suggestion
    /// The response was not as expected, for example the response is an empty vector for
    /// a single query.
```

File: concordium-cis2/src/cis2_client.rs [hunk 27]
```diff
@@ -0,0 +1,696 @@
...
+                mut return_value,
+            } => Ok(Cis2ClientError::InvokeContractError(InvokeContractError::LogicReject {
+                reason,
+                return_value: {
+                    let cis2_error = Cis2Error::<R>::deserial(&mut return_value);
+                    match cis2_error {
+                        Ok(cis2_error) => cis2_error,
+                        Err(_) => bail!(Cis2ClientError::ParseResult),
+                    }
+                },
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:97]:
```suggestion
                return_value: Cis2rror::<R>::deserial(&mut return_value)?
```

File: concordium-cis2/src/cis2_client.rs [hunk 28]
```diff
@@ -0,0 +1,696 @@
...
+                owner,
+                address,
+            }],
+        };
+        let mut res: OperatorOfQueryResponse =
+            self.invoke_contract_read_only(host, OPERATOR_OF_ENTRYPOINT_NAME, params)?;
+        Cis2Client::first(&mut res.0)
+    }
+
+    /// calls the `balanceOf` entrypoint of the CIS2 contract to get the balance
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:223]:
```suggestion
    /// Calls the `balanceOf` entrypoint of the CIS2 contract to get the balance
```

File: concordium-cis2/src/cis2_client.rs [hunk 29]
```diff
@@ -0,0 +1,696 @@
...
+        Ok(res)
+    }
+
+    fn first<T, E>(array: &mut Vec<T>) -> Result<T, Cis2ClientError<E>> {
+        // If the contract returns a response, but the response is empty, it is an
+        // error. Since for a single query the response should be non-empty.
+        ensure!(!array.is_empty(), Cis2ClientError::InvalidResponse);
+
+        Ok(array.swap_remove(0))
+    }
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:395]: This should not be a method, it does not construct `Self` nor does it use it.

Comment (abizjak) [reply concordium-cis2/src/cis2_client.rs:395]: Also, thsi seems to only be used for single queries. In that case it should ensure there is exactly one value in the response.

File: concordium-cis2/src/lib.rs [hunk 30]
```diff
@@ -31,6 +31,9 @@
 //! When `u256_amount` feature is enabled the type [`TokenAmountU256`] is defined
 //! and implements the [`IsTokenAmount`] interface.
 #![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod cis2_client;
```
Comment (abizjak) [inline concordium-cis2/src/lib.rs:35]:
```suggestion
pub use cis2_client::Cis2Client;
```

There is little reason to just expose the submodule.

**Review by limemloh — APPROVED** [review]

File: concordium-cis2/src/cis2_client.rs (hunk 28 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:223]: done

File: concordium-cis2/src/cis2_client.rs [hunk 31]
```diff
@@ -0,0 +1,696 @@
+//! CIS2 client is the intermediatory layer between any contract and
+//! CIS2 comliant contract.
```
[the remark this answers is not in the document]
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:2]: done

File: concordium-cis2/src/cis2_client.rs (hunk 24 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:28]: done

File: concordium-cis2/src/cis2_client.rs (hunk 25 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:30]: done

File: concordium-cis2/src/cis2_client.rs (hunk 26 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:33]: done

File: concordium-cis2/src/cis2_client.rs (hunk 27 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:97]: done, thanks

File: concordium-cis2/src/cis2_client.rs (hunk 29 above)
Comment (parv0888) [reply concordium-cis2/src/cis2_client.rs:395]: removed

File: concordium-cis2/src/lib.rs (hunk 30 above)
Comment (parv0888) [reply concordium-cis2/src/lib.rs:35]: done

**Review by abizjak — APPROVED** [review]

File: concordium-cis2/src/cis2_client.rs [hunk 32]
```diff
@@ -0,0 +1,688 @@
...
+            Err(err) => return Err(Cis2ClientError::<E>::try_from(err)?),
+        };
+
+        Ok(res)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::cis2_client::*;
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:392]:
```suggestion
    use crate::Cis2Client;
```

File: concordium-cis2/src/cis2_client.rs [hunk 33]
```diff
@@ -0,0 +1,688 @@
...
+
+impl<T> From<ParseError> for Cis2ClientError<T> {
+    fn from(_: ParseError) -> Self { Cis2ClientError::ParseResult }
+}
+
+/// Client for interacting with CIS2 compliant contracts.
+///
+/// ## Examples
+/// ```rust
+/// use concordium_cis2::cis2_client::Cis2Client;
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:108]:
```suggestion
/// use concordium_cis2::Cis2Client;
```

File: concordium-cis2/src/cis2_client.rs [hunk 34]
```diff
@@ -0,0 +1,688 @@
...
+    /// If the contract supports CIS2 standard, it returns
+    /// `Ok(SupportResult::Support)`, else it returns
+    /// `Ok(SupportResult::NoSupport)`. If the contract supports CIS2
+    /// standard by another contract, it returns
+    /// `Ok(SupportResult::SupportBy(Vec<ContractAddress>))`. If there is an
+    /// error, it returns `Err`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use concordium_cis2::{cis2_client::*, *};
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:135]:
```suggestion
    /// use concordium_cis2::*;
```

File: concordium-cis2/src/cis2_client.rs [hunk 35]
```diff
@@ -0,0 +1,688 @@
...
+
+    /// Calls the `operatorOf` entrypoint of the CIS2 contract to check if the
+    /// given owner is an operator of the given contract. If the owner is an
+    /// operator of the given contract, it returns `Ok(true)`,
+    /// else it returns `Ok(false)`.
+    /// If there is an error, it returns `Err`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use concordium_cis2::{cis2_client::*, *};
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:179]:
```suggestion
    /// use concordium_cis2::*;
```

File: concordium-cis2/src/cis2_client.rs [hunk 36]
```diff
@@ -0,0 +1,688 @@
...
+
+        Ok(res)
+    }
+
+    /// Calls the `balanceOf` entrypoint of the CIS2 contract to get the balance
+    /// of the given owner for the given token. If the balance is returned,
+    /// it returns `Ok(balance)`, else it returns `Err`.
+    /// # Examples
+    /// ```rust
+    /// use concordium_cis2::{cis2_client::*, *};
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:226]:
```suggestion
    /// use concordium_cis2::*;
```

File: concordium-cis2/src/cis2_client.rs [hunk 37]
```diff
@@ -0,0 +1,688 @@
...
+
+    /// Calls the `transfer` entrypoint of the CIS2 contract to transfer the
+    /// given amount of tokens from the given owner to the given receiver.
+    /// If the transfer is successful and the state is modified, it returns
+    /// `Ok(true)`, else it returns `Ok(false)`. If there is an error, it
+    /// returns `Err`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use concordium_cis2::{cis2_client::*, *};
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:271]:
```suggestion
    /// use concordium_cis2::*;
```

File: concordium-cis2/src/cis2_client.rs [hunk 38]
```diff
@@ -0,0 +1,688 @@
...
+    }
+
+    /// Calls the `updateOperator` of the CIS2 contract.
+    /// If the update is successful and the state is modified, it returns
+    /// `Ok(true)`, else it returns `Ok(false)`. If there is an error, it
+    /// returns `Err`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use concordium_cis2::{cis2_client::*, *};
```
Comment (abizjak) [inline concordium-cis2/src/cis2_client.rs:310]:
```suggestion
    /// use concordium_cis2::*;
```

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
@parv0888 some imports need to be adjusted. Looks good otherwise.

File: concordium-cis2/src/lib.rs [hunk 39]
```diff
@@ -31,6 +31,11 @@
 //! When `u256_amount` feature is enabled the type [`TokenAmountU256`] is defined
 //! and implements the [`IsTokenAmount`] interface.
 #![cfg_attr(not(feature = "std"), no_std)]
+
+mod cis2_client;
+pub use cis2_client::Cis2Client;
+pub use cis2_client::Cis2ClientError;
```
Comment (parv0888) [inline concordium-cis2/src/lib.rs:37]: @abizjak : just to highlight, also had to export `Cis2ClientError`.

Comment (abizjak) [reply concordium-cis2/src/lib.rs:37]: Good catch. Makes sense.

Comment (abizjak) [reply concordium-cis2/src/lib.rs:37]: You probably need to reformat though ;)

**Review by abizjak — APPROVED** [review]

**Pull request merged**